### PR TITLE
[13.0][FIX] product: still missing attributes and relations

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Explained in code, see each commit. It's a continuation of https://github.com/OCA/OpenUpgrade/pull/2633.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I don't see enough explanations neither on the PR message nor the commit messages. Please do the same as I did in #2633.

**Comment by MiquelRForgeFlow** [comment]
Yeah, yesterday I did it a bit fast.

The second commit is easy to explain. After migration of a client that has mrp, there were some bom lines that had lost their attribute, just the same way/problem you had previously with products that you solved previously. And example for this case should be a copy pasta of the ones you put but changing products by bom lines. And that's why the queries are very similar.

About the first commit it's a bit tricky. It happens that the new archived attributes you created, after migration, they still had null attribute_line_id, which is a required field. The problem lies in that when you created the archived attribute lines, you should also fill the m2m table between PAV and PTAL, and that's what the commit basically does. It also includes two things: 1) in postmigration, first check for active ptals and then for inactive ptals, always trying to find ptals because we want all attribute_line_id filled. 2) in pre-migration, we begin to fill attribute_line_id for the new archived ptavs (using the previous m2m table between PAV and PTAL), and thus in post-migration this ones won't need have attribute_line_id filled.

**Comment by MiquelRForgeFlow** [comment]
> I don't see enough explanations neither on the PR message nor the commit messages

Done. Now you can check commit by commit what each minor change is about.

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (pedrobaeza) [inline addons/product/migrations/13.0.1.2/pre-migration.py:179]: If the ptvas were archived on v13, they has combination_indices, isn't it? Why not filling such ones then?

File: addons/product/migrations/13.0.1.2/openupgrade_analysis_work.txt
Comment (pedrobaeza) [inline addons/product/migrations/13.0.1.2/openupgrade_analysis_work.txt:86]: If I don't remember bad, if you this, there are missing values that are fill lately.

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (pedrobaeza) [inline addons/product/migrations/13.0.1.2/pre-migration.py]: Please don't nest this high level methods.

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (pedrobaeza) [inline addons/product/migrations/13.0.1.2/pre-migration.py:75]: I don't get the relation between product_template_attribute_line and mrp boms, neither the code has any reference to this.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
I still miss an explanation of each change about why, and enumerating the steps to reproduce the problem. Please take as example https://github.com/OCA/OpenUpgrade/pull/2633, where the exact initial data and the expected ones are shown.

Without that, everything is done blindly, and the SQL modifications are very convoluted for validating them reading them.

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:179]: Ok, you have a point here. I don't know why but I thought archived ptavs were avoided. Dropped that commit.

File: addons/product/migrations/13.0.1.2/openupgrade_analysis_work.txt
Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/openupgrade_analysis_work.txt:86]: I initially moved that to pre-migration because the field attribute_line_id of ptav is required. Later, it proves useful in pre-migration, in correctly filling the missing ptavs (the query of missing ptavs use that field in order to make it compatible with the sql_constraint `unique(attribute_line_id, product_attribute_value_id)`)

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py]: Ok, done.

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:75]: It's explained in that commit message. In mrp module, the bom lines also have attributes, and some are missing in a similar way the missing attributes on products case you detected. In order to avoid these missing attributes being missing, we need to make use of mrp_bom_line_product_attribute_value_rel, which plays a very similar role that product_attribute_value_product_product_rel has for products. We have tested it and works excellent.

**Comment by MiquelRForgeFlow** [comment]
> I still miss an explanation of each change about why, and enumerating the steps to reproduce the problem

Well, I thought I added the enough explanations in the commit message, in the code or in the work.txt (depending the commit). Just tell me exactly which commits you need more info about and I will try to add more explanation on them.

**Comment by MiquelRForgeFlow** [comment]
PR updated :)

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (kos94ok-3D) [inline addons/product/migrations/13.0.1.2/pre-migration.py]: Now we can do it in one query:
<pre>
  openupgrade.logged_query(
      env.cr, """
      UPDATE product_template_attribute_value ptav
      SET attribute_line_id = ptal.id
      FROM product_template_attribute_line ptal
      JOIN product_attribute_value pav ON ptal.attribute_id = pav.attribute_id
      WHERE ptav.product_tmpl_id = ptal.product_tmpl_id
          AND ptav.product_attribute_value_id = pav.id
          AND ptav.attribute_line_id IS NULL""",
  )
</pre>

File: addons/mrp/migrations/13.0.2.0/post-migration.py
Comment (kos94ok-3D) [inline addons/mrp/migrations/13.0.2.0/post-migration.py]: useless join we need only `id` which we already have in `mb.product_tmpl_id`

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (kos94ok-3D) [inline addons/product/migrations/13.0.1.2/pre-migration.py]: useless join we need only `id` which we already have in `mb.product_tmpl_id`

File: addons/product/migrations/13.0.1.2/pre-migration.py
Comment (kos94ok-3D) [inline addons/product/migrations/13.0.1.2/pre-migration.py]: useless join we need only `id` which we already have in `mb.product_tmpl_id`

**Review by kos94ok-3D — COMMENTED** [review]
Work as expected.
I have some cosmetic suggests

File: addons/mrp/migrations/13.0.2.0/post-migration.py
Comment (MiquelRForgeFlow) [reply addons/mrp/migrations/13.0.2.0/post-migration.py]: that's https://github.com/OCA/OpenUpgrade/pull/2957 😆

**Comment by kos94ok-3D** [comment]
@MiquelRForgeFlow You forgot about `group by` section

**Comment by MiquelRForgeFlow** [comment]
@kos94ok-3D done

**Review by kos94ok-3D — APPROVED** [review]

**Comment by kos94ok-3D** [comment]
HI, @MiquelRForgeFlow
Are we waiting for the review from @pedrobaeza?

**Comment by MiquelRForgeFlow** [comment]
> Are we waiting for the review from @pedrobaeza?

Yeah, a long awaited review. He wants me to wait for his reviews in this repo for all the PRs.

**Comment by pedrobaeza** [comment]
I still need commit messages explaining the whys of these commits:

97142fbe68e913dc245225828c533887c2326060
c263c885cb9ff813d7336046b4af3f2a7e535ee2
e54a1412c6c2710973a4783eb148cafbb8759e20
1bc8c9820c6dbec4e277cdc00114579264da409e
f6dbddc5400cc22fbf084e743cc7f42c9e5eeac0
a823cdec234b7b6902349f147b6f858f987cafb9

and steps to reproduce a problem.

Please see a good example like the one I did for a similar case: 3dabbccb7d96c9627fae0f1e4fbdf790c1589be9

**Comment by MiquelRForgeFlow** [comment]
- https://github.com/OCA/OpenUpgrade/commit/97142fbe68e913dc245225828c533887c2326060 - it's just a simplification. It reduces the selects by one.
- https://github.com/OCA/OpenUpgrade/commit/e54a1412c6c2710973a4783eb148cafbb8759e20 is the core of this PR. All this refactor began with this commit, which fixes the correct number of archived ptavs that are being created. The key is adding this condition in the left join:
Selection_054
And to add this condition, then we need to have that `ptav.attribune_line_id` filled, which is done previously in https://github.com/OCA/OpenUpgrade/commit/c263c885cb9ff813d7336046b4af3f2a7e535ee2.
- https://github.com/OCA/OpenUpgrade/commit/1bc8c9820c6dbec4e277cdc00114579264da409e already has the explanation included in the docstring (of `fill_product_template_attribute_value__attribute_id_related` method).
- https://github.com/OCA/OpenUpgrade/commit/f6dbddc5400cc22fbf084e743cc7f42c9e5eeac0 just solves a similar case of losing data (and thus not having enough archived ptavs). The mrp_bom_line_product_attribute_value_rel table becomes deprecated and is replaced by mrp_bom_line_product_template_attribute_value_rel in mrp module. This extra archived ptavs are created here instead of doing it in mrp module because we can take advantage of the other queries to fill the collateral data. If not, in mrp we would have to copy and paste the same queries, like the one that fills attribute_id in ptavs, and thus duplicating a lot of code.
- https://github.com/OCA/OpenUpgrade/commit/a823cdec234b7b6902349f147b6f858f987cafb9 this one is a pretty obvious fix and I doubt it needs any explanation. Again, without this extra conditions, we are having wrong data.

**Comment by kos94ok-3D** [comment]
Case1 Steps to reproduce:
1. Create a Product Template with
   Attribute lines: `Color (Red, Blue)`, `Material (Iron, Wood)`
 2. Create and Validate transfer with Variant `(Red, Iron)`
 3. Remove `Red` value from attribute line
  (on this step we lose here link between `product.attribute.value` and `product.template.attribute.value`)
 4. run migration
 5. Go to Product Template form
 6. Press on `Configure Variants` button 
 7. Set `Inactive` filter
 
Actual Result:
Missing attribute for inactive `product.template.attribute.value`

Case 2 Steps to reproduce:
1. Create a Product Template with
   Attribute lines: `Color (Red, Blue)`, `Material (Iron, Wood)`
 2. Create BOM for this Product in the BOM line set apply on `Color: Red`
 3. Remove `Red` value from attribute line
  (on this step `product.template.attribute.value` and `product.variant` will be removed)
 4. run migration
 5. Go to Product Template form
 6. Press on `Configure Variants` button 
 7. Set `Inactive` filter not found 
Actual Result:
Missing `Color: Red` record

8. Go to BOM form
Actual Result:
Missing `Color: Red` in BOM line

**Review by pedrobaeza — APPROVED** [review]
Pushed a rebased branch with all of these comments in commit messages and review thoroughly. There are few doubts about some extra linking, but let's see applied to real migrations.

**Pull request merged**
